{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VIVA: Visualization of Variants\n",
    "\n",
    "### VIVA is a tool for visualizing coverage and variant genotype information from variant call format (VCF) files. \n",
    "\n",
    "Enter your VCF filename below and choose your plotting options.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# @hidden_cell\n",
    "println(\"Welcome to VIVA Jupyter Notebook.\")\n",
    "println()\n",
    "println(\"Loading packages:\")\n",
    "println()\n",
    "\n",
    "println(\"1/5 ArgParse\")\n",
    "using ArgParse\n",
    "\n",
    "println(\"2/5 VariantVisualization\")\n",
    "using Pkg #remove after VariantVisualization is deployed\n",
    "Pkg.activate(@__DIR__) #remove after VariantVisualization is deployed\n",
    "using VariantVisualization\n",
    "\n",
    "println(\"3/5 PlotlyJS\")\n",
    "using PlotlyJS\n",
    "\n",
    "println(\"4/5 GeneticVariation\")\n",
    "using GeneticVariation\n",
    "\n",
    "println(\"5/5 DelimitedFiles\")\n",
    "using DelimitedFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Enter You VCF filename\n",
    "\n",
    "Enter filename or file path of a VCF file to visualize. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_filename = \"/Users/George/.julia/dev/VIVA/test/test_files/test_4X_191.vcf\" ;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Variant Selection\n",
    "\n",
    "Select variant record rows of your VCF file to visualize.\n",
    "\n",
    "*Important Note* Variant selection is recommending when visualizing large VCF files as plotting capability is limited by user's computing power."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomic_range = \"\" #specify genomic range of variants to select in the format chr4:0-560000000\n",
    "pass_filter = \"\" #if \"true\", selects variants with Filter Status = PASS\n",
    "positions_list = \"\" #include path and filename to list of variant positions to selects ;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Sample Grouping and Selection\n",
    "\n",
    "Add sample metadata to heatmap plots and group sample columns by metadata traits. This feature is useful for identifying batch effects or differential variation between samples. \n",
    "\n",
    "Additionally, you may select unique samples to include in heatmap and average read depth scatter plots. \n",
    "\n",
    "*Note* If you would like to group samples using a metadata table in combination with the sample selection feature, the metadata table must only include metadata for the samples to be selected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "group_samples = \"/Users/George/.julia/dev/VIVA/test/test_files/sample_metadata_matrix.csv\" #include path and filename to matrix to group by\n",
    "select_samples = \"\" #include path and filename to list of sample names\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Plotting Options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmaps = \"genotype,read_depth\" #make default both - genotype, read_depth, or genotype,read_depth\n",
    "read_depth_scatter_plots = \"\" \n",
    "y_axis_label_option = \"chromosomes\"#defaults to \"chromosome labels\". If \"positions\", each variant position is labeled on y axis. If \"none\", no y-axis ticklabels are plotted.\n",
    "x_axis_label_option = \"\" #if true, sample ids are labelled on x axis labels\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Saving Options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap_title = \"\"\n",
    "save_format = \"html\"\n",
    "output_directory = \"\"\n",
    "num_array = \"\" #if \"true\", saves labeled number array of filtered heatmap input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# @hidden_cell\n",
    "\n",
    "#Passing settings to VIVA. Do not adjust.\n",
    "\n",
    "variant_filters = genomic_range,pass_filter,positions_list\n",
    "sample_selection = group_samples,select_samples\n",
    "plotting_options = heatmap,read_depth_scatter_plots,y_axis_label_option,x_axis_label_option\n",
    "saving_options = heatmap_title,save_format,output_directory,num_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# @hidden_cell\n",
    "println(\"Welcome to VIVA Jupyter Notebook.\")\n",
    "println()\n",
    "println(\"Loading packages:\")\n",
    "println()\n",
    "\n",
    "println(\"1/5 ArgParse\")\n",
    "using ArgParse\n",
    "\n",
    "println(\"2/5 VIVA\")\n",
    "using Pkg #remove after VIVA is deployed\n",
    "Pkg.activate(\"/Users/George/.julia/dev/VIVA\") #remove after VIVA is deployed\n",
    "using VIVA\n",
    "\n",
    "println(\"3/5 PlotlyJS\")\n",
    "using PlotlyJS\n",
    "\n",
    "println(\"4/5 GeneticVariation\")\n",
    "using GeneticVariation\n",
    "\n",
    "println(\"5/5 DelimitedFiles\")\n",
    "using DelimitedFiles\n",
    "\n",
    "\"\"\"\n",
    "jupyter_main(vcf_filename,saving_options,variant_filters,sample_selection,plotting_options)\n",
    "\n",
    "filters, plots visualization, and saves as figure.\n",
    "utilizes all global variables set in first cell of jupyter notebook\n",
    "\"\"\"\n",
    "function jupyter_main(vcf_filename,saving_options,variant_filters,sample_selection,plotting_options)\n",
    "\n",
    "    genomic_range=variant_filters[1]\n",
    "    pass_filter=variant_filters[2]\n",
    "    positions_list=variant_filters[3]\n",
    "    group_samples=sample_selection[1]\n",
    "    select_samples=sample_selection[2]\n",
    "    heatmap=plotting_options[1]\n",
    "    read_depth_scatter_plots=plotting_options[2]\n",
    "    y_axis_label_option=plotting_options[3]\n",
    "    x_axis_labels=plotting_options[4]\n",
    "    heatmap_title=saving_options[1]\n",
    "    save_format=saving_options[2]\n",
    "    output_directory=saving_options[3]\n",
    "    num_array=saving_options[4]\n",
    "\n",
    "    #create vcf reader object\n",
    "    println(\"Reading $vcf_filename\")\n",
    "    reader = VCF.Reader(open(vcf_filename, \"r\"))\n",
    "\n",
    "    #make sample names object\n",
    "    sample_names = get_sample_names(reader)\n",
    "\n",
    "    #store save format and create output directory if it doesn't exist yet\n",
    "#=\n",
    "    save_ext = save_format\n",
    "    VIVA.checkfor_outputdirectory(output_directory)\n",
    "    output_directory=output_directory\n",
    "=#\n",
    "\n",
    "    save_ext = save_format\n",
    "    output_directory=output_directory\n",
    "\n",
    "    #store plot label options\n",
    "    if occursin(\"true\",x_axis_labels)\n",
    "        x_axis_label_option = true\n",
    "    else\n",
    "        x_axis_label_option = false\n",
    "    end\n",
    "\n",
    "    println(y_axis_label_option)\n",
    "\n",
    "    #check for filters and apply then show stats again\n",
    "\n",
    "    #no filter\n",
    "    if !occursin(\".csv\",positions_list) && !occursin(\"chr\",genomic_range) && !occursin(\"true\",pass_filter)\n",
    "        println(\"no filters applied. Large vcf files will take a long time to process and heatmap visualizations will not be useful at this scale.\")\n",
    "        println()\n",
    "        println(\"Loading VCF file into memory for visualization\")\n",
    "\n",
    "        sub = Array{Any}(undef,0)\n",
    "        for record in reader\n",
    "            push!(sub,record)\n",
    "        end\n",
    "    end\n",
    "\n",
    "    #all filter combos\n",
    "\n",
    "    #pass_filter and genomic_range and list\n",
    "    if occursin(\"true\",pass_filter) && occursin(\"chr\",genomic_range) && occursin(\".csv\",positions_list)\n",
    "        sig_list =  load_siglist(positions_list)\n",
    "        sub = VIVA.pass_chrrange_siglist_filter(vcf_filename, sig_list, genomic_range)\n",
    "        number_rows = size(sub,1)\n",
    "        println(\"Selected $number_rows variants with Filter status: PASS, that match list of chromosome positions of interest, and are within chromosome range: $genomic_range\")\n",
    "    end\n",
    "\n",
    "    #pass_filter and genomic_range\n",
    "    if occursin(\"true\",pass_filter) && occursin(\"chr\",genomic_range)\n",
    "        sub = VIVA.pass_chrrange_filter(reader, genomic_range,vcf_filename)\n",
    "        number_rows = size(sub,1)\n",
    "        println(\"Selected $number_rows variants with Filter status: PASS and are within chromosome range: $genomic_range\")\n",
    "    end\n",
    "\n",
    "    #pass_filter and list\n",
    "    if occursin(\"true\",pass_filter) && occursin(\".csv\",positions_list)\n",
    "        sig_list =  load_siglist(positions_list)\n",
    "        sub = VIVA.pass_siglist_filter(vcf_filename, sig_list)\n",
    "        number_rows = size(sub,1)\n",
    "        println(\"Selected $number_rows variants with Filter status: PASS and that match list of chromosome positions of interest\")\n",
    "    end\n",
    "\n",
    "    #genomic_range and list\n",
    "    if occursin(\"chr\",genomic_range) && occursin(\".csv\",positions_list)\n",
    "        sig_list =  load_siglist(positions_list)\n",
    "        sub = VIVA.chrrange_siglist_filter(vcf_filename, sig_list, genomic_range)\n",
    "        number_rows = size(sub,1)\n",
    "        println(\"Selected $number_rows variants that are within chromosome range: $genomic_range and that match list of chromosome positions of interest\")\n",
    "    end\n",
    "\n",
    "    #pass_filter\n",
    "    if occursin(\"true\",pass_filter) && !occursin(\"chr\",genomic_range) && !occursin(\".csv\",positions_list)\n",
    "        println(\"Only pass filter is applied. Large vcf files with many PASS variants will take a long time to process and heatmap visualizations will lose resolution at this scale unless viewed in interactive html for zooming.\")\n",
    "        println()\n",
    "        sub = VIVA.io_pass_filter(vcf_filename)\n",
    "        number_rows = size(sub,1)\n",
    "        println(\"Selected $number_rows variants with Filter status: PASS\")\n",
    "        heatmap_input = \"pass_filtered\"\n",
    "    end\n",
    "\n",
    "    #genomic_range\n",
    "    if occursin(\"chr\",genomic_range) && !occursin(\"true\",pass_filter) && !occursin(\".csv\",positions_list)\n",
    "        sub = VIVA.io_chromosome_range_vcf_filter(genomic_range,vcf_filename)\n",
    "        number_rows = size(sub,1)\n",
    "        println(\"Selected $number_rows variants within chromosome range: $genomic_range\")\n",
    "        heatmap_input = \"range_filtered\"\n",
    "    end\n",
    "\n",
    "    #list\n",
    "    if occursin(\".csv\",positions_list) && !occursin(\"chr\",genomic_range) && !occursin(\"true\",pass_filter)\n",
    "        sig_list =  load_siglist(positions_list)\n",
    "        sub = VIVA.io_sig_list_vcf_filter(sig_list,vcf_filename)\n",
    "        number_rows = size(sub,1)\n",
    "        println(\"Selected $number_rows variants that match list of chromosome positions of interest\")\n",
    "        heatmap_input = \"positions_filtered\"\n",
    "    end\n",
    "\n",
    "    println()\n",
    "    println(\"Finished Filtering. Total time to filter:\")\n",
    "    println(\"_______________________________________________\")\n",
    "\n",
    "\n",
    "    if occursin(\"genotype\",heatmap) && !occursin(\"read_depth\",heatmap)\n",
    "        gt_num_array,gt_chromosome_labels = combined_all_genotype_array_functions(sub)\n",
    "\n",
    "        if heatmap_title != \"\"\n",
    "            title = \"Genotype_$(heatmap_title)\"\n",
    "        else\n",
    "            bn = Base.Filesystem.basename(vcf_filename)\n",
    "            title = \"Genotype_$bn\"\n",
    "        end\n",
    "\n",
    "        chr_pos_tuple_list = generate_chromosome_positions_for_hover_labels(gt_chromosome_labels)\n",
    "\n",
    "        chrom_label_info = VIVA.chromosome_label_generator(gt_chromosome_labels[:,1])\n",
    "\n",
    "        if length(split(group_samples,\",\")) == 2\n",
    "\n",
    "            if select_samples != \"\"\n",
    "\n",
    "                println(\"Make sure that select_samples list and phenotype matrix contain the same sample names.\")\n",
    "                println(\"Selecting samples listed in $select_samples\")\n",
    "\n",
    "                gt_num_array,col_selectedcolumns = select_columns(select_samples,\n",
    "                                              gt_num_array,\n",
    "                                              sample_names)\n",
    "\n",
    "                sample_names = col_selectedcolumns\n",
    "            end\n",
    "\n",
    "            group_trait_matrix_filename=(split(group_samples,\",\")[1])\n",
    "            trait_to_group_by = (split(group_samples,\",\")[2])\n",
    "            println()\n",
    "            println(\"Grouping samples by $trait_to_group_by\")\n",
    "            println()\n",
    "\n",
    "            ordered_num_array,group_label_pack,pheno,id_list,trait_labels = sortcols_by_phenotype_matrix(group_trait_matrix_filename, trait_to_group_by, gt_num_array, sample_names)\n",
    "\n",
    "            if occursin(\"true\",num_array)\n",
    "            save_numerical_array(ordered_num_array,sample_names,chr_pos_tuple_list,title,output_directory)\n",
    "            end\n",
    "\n",
    "            pheno_num_array,trait_label_array,chrom_label_info=add_pheno_matrix_to_gt_data_for_plotting(ordered_num_array,pheno,trait_labels,chrom_label_info,number_rows)\n",
    "\n",
    "            graphic = VIVA.genotype_heatmap_with_groups(pheno_num_array,title,chrom_label_info,group_label_pack,id_list,chr_pos_tuple_list,y_axis_label_option,trait_label_array,x_axis_label_option,number_rows)\n",
    "             return graphic\n",
    "        else\n",
    "\n",
    "            if select_samples != \"\"\n",
    "\n",
    "                println(\"Selecting samples listed in $select_samples\")\n",
    "\n",
    "                gt_num_array,col_selectedcolumns = select_columns(select_samples,\n",
    "                                              gt_num_array,\n",
    "                                              sample_names)\n",
    "\n",
    "                sample_names = col_selectedcolumns\n",
    "            end\n",
    "\n",
    "            if occursin(\"true\",num_array)\n",
    "            save_numerical_array(gt_num_array,sample_names,chr_pos_tuple_list,title,output_directory)\n",
    "            end\n",
    "\n",
    "            graphic = VIVA.genotype_heatmap2(gt_num_array,title,chrom_label_info,sample_names,chr_pos_tuple_list,y_axis_label_option,x_axis_label_option)\n",
    "             return graphic\n",
    "            \n",
    "        end\n",
    "\n",
    "        println(\"Saving genotype heatmap\")\n",
    "\n",
    "        save_graphic(graphic,output_directory,save_ext,title)\n",
    "\n",
    "    end\n",
    "\n",
    "    if occursin(\"read_depth\",heatmap) && !occursin(\"genotype\",heatmap)\n",
    "\n",
    "        dp_num_array,dp_chromosome_labels = combined_all_read_depth_array_functions(sub)\n",
    "\n",
    "        if heatmap_title != \"\"\n",
    "            title = \"Read_Depth_$heatmap_title\"\n",
    "        else\n",
    "            bn = Base.Filesystem.basename(vcf_filename)\n",
    "            title = \"Read_Depth_$bn\"\n",
    "        end\n",
    "\n",
    "        chr_pos_tuple_list = generate_chromosome_positions_for_hover_labels(dp_chromosome_labels)\n",
    "\n",
    "        chrom_label_info = VIVA.chromosome_label_generator(dp_chromosome_labels[:,1])\n",
    "\n",
    "        if length(split(group_samples,\",\")) == 2\n",
    "\n",
    "            if select_samples != \"\"\n",
    "\n",
    "                println(\"Make sure that select_samples list and phenotype matrix contain the same sample names.\")\n",
    "                println(\"Selecting samples listed in $select_samples\")\n",
    "\n",
    "                gt_num_array,col_selectedcolumns = select_columns(select_samples,\n",
    "                                              dp_num_array,\n",
    "                                              sample_names)\n",
    "\n",
    "                sample_names = col_selectedcolumns\n",
    "            end\n",
    "\n",
    "\n",
    "            group_trait_matrix_filename=(split(group_samples,\",\")[1])\n",
    "            trait_to_group_by = (split(group_samples,\",\")[2])\n",
    "            println()\n",
    "            println(\"Grouping samples by $trait_to_group_by\")\n",
    "            println()\n",
    "\n",
    "            ordered_dp_num_array,group_label_pack,pheno,id_list,trait_labels = sortcols_by_phenotype_matrix(group_trait_matrix_filename, trait_to_group_by, dp_num_array, sample_names)\n",
    "            dp_num_array_limited=read_depth_threshhold(ordered_dp_num_array)\n",
    "\n",
    "             if occursin(\"true\",num_array)\n",
    "            save_numerical_array(ordered_dp_num_array,sample_names,chr_pos_tuple_list,title,output_directory)\n",
    "            end\n",
    "\n",
    "            pheno_num_array,trait_label_array,chrom_label_info = add_pheno_matrix_to_dp_data_for_plotting(dp_num_array_limited,pheno,trait_labels,chrom_label_info,number_rows)\n",
    "\n",
    "            graphic = VIVA.dp_heatmap2_with_groups(pheno_num_array,title,chrom_label_info,group_label_pack,id_list,chr_pos_tuple_list,y_axis_label_option,trait_label_array,x_axis_label_option,number_rows)\n",
    "            return graphic\n",
    "            \n",
    "        else\n",
    "\n",
    "               if select_samples != \"\"\n",
    "\n",
    "                   println(\"Selecting samples listed in $select_samples\")\n",
    "\n",
    "                   dp_num_array,col_selectedcolumns = select_columns(select_samples,\n",
    "                                                 gt_num_array,\n",
    "                                                 sample_names)\n",
    "\n",
    "                   sample_names = col_selectedcolumns\n",
    "               end\n",
    "\n",
    "            if occursin(\"true\",num_array)\n",
    "            save_numerical_array(dp_num_array,sample_names,chr_pos_tuple_list,title,output_directory)\n",
    "            end\n",
    "\n",
    "            dp_num_array_limited=read_depth_threshhold(dp_num_array)\n",
    "\n",
    "            graphic = VIVA.dp_heatmap2(dp_num_array, title, chrom_label_info, sample_names,chr_pos_tuple_list,y_axis_label_option,x_axis_label_option)\n",
    "           return graphic\n",
    "        \n",
    "        end\n",
    "\n",
    "        println(\"Saving read depth heatmap\")\n",
    "\n",
    "        save_graphic(graphic,output_directory,save_ext,title)\n",
    "    end\n",
    "\n",
    "    if occursin(\"read_depth\",heatmap) && occursin(\"genotype\",heatmap)\n",
    "        \n",
    "        println(\"both heatmaps\")\n",
    "\n",
    "        gt_num_array,gt_chromosome_labels = combined_all_genotype_array_functions(sub)\n",
    "\n",
    "        if heatmap_title != \"\"\n",
    "            title = \"Genotype_$(heatmap_title)\"\n",
    "        else\n",
    "            bn = Base.Filesystem.basename(vcf_filename)\n",
    "            title = \"Genotype_$bn\"\n",
    "        end\n",
    "\n",
    "        chr_pos_tuple_list = generate_chromosome_positions_for_hover_labels(gt_chromosome_labels)\n",
    "\n",
    "        chrom_label_info = VIVA.chromosome_label_generator(gt_chromosome_labels[:,1])\n",
    "\n",
    "        if length(split(group_samples,\",\")) == 2\n",
    "\n",
    "            if select_samples != \"\"\n",
    "\n",
    "                println(\"Make sure that select_samples list and phenotype matrix contain the same sample names.\")\n",
    "                println(\"Selecting samples listed in $select_samples\")\n",
    "\n",
    "                gt_num_array,col_selectedcolumns = select_columns(select_samples,\n",
    "                                              gt_num_array,\n",
    "                                              sample_names)\n",
    "\n",
    "                sample_names = col_selectedcolumns\n",
    "            end\n",
    "\n",
    "            group_trait_matrix_filename=(split(group_samples,\",\")[1])\n",
    "            trait_to_group_by = (split(group_samples,\",\")[2])\n",
    "            println()\n",
    "            println(\"Grouping samples by $trait_to_group_by\")\n",
    "            println()\n",
    "\n",
    "            ordered_num_array,group_label_pack,pheno,id_list,trait_labels = sortcols_by_phenotype_matrix(group_trait_matrix_filename, trait_to_group_by, gt_num_array, sample_names)\n",
    "\n",
    "            if occursin(\"true\",num_array)\n",
    "            save_numerical_array(ordered_num_array,sample_names,chr_pos_tuple_list,title,output_directory)\n",
    "            end\n",
    "\n",
    "            pheno_num_array,trait_label_array,chrom_label_info=add_pheno_matrix_to_gt_data_for_plotting(ordered_num_array,pheno,trait_labels,chrom_label_info,number_rows)\n",
    "\n",
    "            graphic = VIVA.genotype_heatmap_with_groups(pheno_num_array,title,chrom_label_info,group_label_pack,id_list,chr_pos_tuple_list,y_axis_label_option,trait_label_array,x_axis_label_option,number_rows)\n",
    "            \n",
    "            \n",
    "        else\n",
    "\n",
    "            if select_samples != \"\"\n",
    "\n",
    "                println(\"Selecting samples listed in $select_samples\")\n",
    "\n",
    "                gt_num_array,col_selectedcolumns = select_columns(select_samples,\n",
    "                                              gt_num_array,\n",
    "                                              sample_names)\n",
    "\n",
    "                sample_names = col_selectedcolumns\n",
    "            end\n",
    "\n",
    "            if occursin(\"true\",num_array)\n",
    "            save_numerical_array(gt_num_array,sample_names,chr_pos_tuple_list,title,output_directory)\n",
    "            end\n",
    "\n",
    "            graphic = VIVA.genotype_heatmap2(gt_num_array,title,chrom_label_info,sample_names,chr_pos_tuple_list,y_axis_label_option,x_axis_label_option)\n",
    "            \n",
    "            \n",
    "        end\n",
    "\n",
    "        println(\"Saving genotype heatmap\")\n",
    "\n",
    "        save_graphic(graphic,output_directory,save_ext,title)\n",
    "\n",
    "        if heatmap_title != \"\"\n",
    "            title = \"Read_Depth_$heatmap_title\"\n",
    "        else\n",
    "            bn = Base.Filesystem.basename(vcf_filename)\n",
    "            title = \"Read_Depth_$bn\"\n",
    "        end\n",
    "\n",
    "        dp_num_array,dp_chromosome_labels = combined_all_read_depth_array_functions(sub)\n",
    "\n",
    "        chr_pos_tuple_list = generate_chromosome_positions_for_hover_labels(dp_chromosome_labels)\n",
    "\n",
    "        chrom_label_info = VIVA.chromosome_label_generator(dp_chromosome_labels[:,1])\n",
    "\n",
    "        if length(split(group_samples,\",\")) == 2\n",
    "\n",
    "            if select_samples != \"\"\n",
    "\n",
    "                println(\"Make sure that select_samples list and phenotype matrix contain the same sample names.\")\n",
    "                println(\"Selecting samples listed in $select_samples\")\n",
    "\n",
    "                gt_num_array,col_selectedcolumns = select_columns(select_samples,\n",
    "                                              dp_num_array,\n",
    "                                              sample_names)\n",
    "\n",
    "                sample_names = col_selectedcolumns\n",
    "            end\n",
    "\n",
    "            group_trait_matrix_filename=(split(group_samples,\",\")[1])\n",
    "            trait_to_group_by = (split(group_samples,\",\")[2])\n",
    "            println()\n",
    "            println(\"Grouping samples by $trait_to_group_by\")\n",
    "            println()\n",
    "\n",
    "            ordered_dp_num_array,group_label_pack,pheno,id_list,trait_labels = sortcols_by_phenotype_matrix(group_trait_matrix_filename, trait_to_group_by, dp_num_array, sample_names)\n",
    "            dp_num_array_limited=read_depth_threshhold(ordered_dp_num_array)\n",
    "\n",
    "            if occursin(\"true\",num_array)\n",
    "            save_numerical_array(ordered_dp_num_array,sample_names,chr_pos_tuple_list,title,output_directory)\n",
    "            end\n",
    "\n",
    "            pheno_num_array,trait_label_array,chrom_label_info = add_pheno_matrix_to_dp_data_for_plotting(dp_num_array_limited,pheno,trait_labels,chrom_label_info,number_rows)\n",
    "\n",
    "            graphic = VIVA.dp_heatmap2_with_groups(pheno_num_array,title,chrom_label_info,group_label_pack,id_list,chr_pos_tuple_list,y_axis_label_option,trait_label_array,x_axis_label_option,number_rows)\n",
    "            \n",
    "            \n",
    "        else\n",
    "\n",
    "\n",
    "           if select_samples != \"\"\n",
    "\n",
    "               println(\"Selecting samples listed in $select_samples\")\n",
    "\n",
    "               dp_num_array,col_selectedcolumns = select_columns(select_samples,\n",
    "                                             gt_num_array,\n",
    "                                             sample_names)\n",
    "\n",
    "               sample_names = col_selectedcolumns\n",
    "           end\n",
    "\n",
    "            if occursin(\"true\",num_array)\n",
    "            save_numerical_array(dp_num_array,sample_names,chr_pos_tuple_list,title,output_directory)\n",
    "            end\n",
    "\n",
    "            dp_num_array_limited=read_depth_threshhold(dp_num_array)\n",
    "            \n",
    "            graphic = VIVA.dp_heatmap2(dp_num_array, title, chrom_label_info, sample_names,chr_pos_tuple_list,y_axis_label_option,x_axis_label_option)\n",
    "         \n",
    "            \n",
    "        end\n",
    "\n",
    "        println(\"Saving read depth heatmap\")\n",
    "\n",
    "        save_graphic(graphic,output_directory,save_ext,title)\n",
    "        \n",
    "        \n",
    "    end\n",
    "    \n",
    "    \n",
    "end\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "jupyter_main(vcf_filename,saving_options,variant_filters,sample_selection,plotting_options)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 0.7.0",
   "language": "julia",
   "name": "julia-0.7"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "0.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
